### PR TITLE
BE-8005: before and after

### PR DESCRIPTION
**Pull request by ajanata**
some other minor cleanup

added lots of tests, fixed a couple bugs the tests found

### CODE REVIEW

File: orderedmap/map_test.go [hunk 1]
```diff
@@ -1,10 +1,10 @@
 package orderedmap
 
 import (
-	"testing"
+	"fmt"
 )
 
-func TestOrderedMap(t *testing.T) {
+func ExampleOrderedMap() {
```
Comment (ajanata) [inline orderedmap/map_test.go:7]: Examples are run as tests and will fail if the output does not match the provided output in the comments.

File: register.go [hunk 2]
```diff
@@ -32,25 +23,158 @@ func RegisterTest(name string, tester Tester) any {
...
+//
+// For more details, see https://go.dev/ref/spec#Package_initialization
+
+// Test registers a new test to be run. Tests are run in lexicographical order within a package.
+// TODO: support the -shuffle testflag. though maybe that only has to be in the runner?
+//
+// The return value may be discarded (and is always nil); it is provided to simplify writing test code, like so:
+//
+//    var _ = testy.Test("my test", func(t testy.TestingT){})
+func Test(name string, tester Tester) any {
```
Comment (ajanata) [inline register.go:60]: Revisiting the name on this. Propose renaming this to align with `BeforePackage` etc.; `RegisterBeforePackage` is way too wordy.

File: register.go [hunk 3]
```diff
@@ -32,25 +23,158 @@ func RegisterTest(name string, tester Tester) any {
...
 		Package: pkg,
 		Name:    name,
 		tester:  tester,
 	}
 
 	return nil
 }
+
+// BeforePackage registers a function to be run once before any tests in the given package are run. A package may only
+// have one BeforePackage function.
```
Comment (ajanata) [inline register.go:83]: This (and the others) could be expanded to allow multiple helpers, if that proves useful, but an execution order would have to be defined (likely the order that initialization occurs). That also can be done without breaking existing users, so I see no reason to worry about it until it does actually show itself to be useful.

File: orderedmap/map.go [hunk 4]
```diff
@@ -1,8 +1,9 @@
 package orderedmap
 
 import (
-	"constraints"
 	"sort"
+
+	"golang.org/x/exp/constraints"
```
Comment (ajanata) [inline orderedmap/map.go:6]: `constraints` is going to be removed from the standard library in the next beta/rc build, but it has already been pushed to experimental, so making this change now

File: run.go [hunk 5]
```diff
@@ -13,48 +15,223 @@ import (
...
+// TODO: shuffle test execution order (see -shuffle in `go help testflag`)
 func RunAsTest(t *testing.T) {
 	t.Helper()
-	instance.tests.Iterate(func(pkg string, tests orderedmap.OrderedMap[string, testCase]) bool {
-		tests.Iterate(func(name string, test testCase) bool {
-			t.Run(test.Name, func(tt *testing.T) {
-				tt.Helper()
-				test.tester(tWrapper{t: tt})
+	instance.tests.Iterate(func(pkg string, pkgTests *testPkg) bool {
+		// we have to hold onto any panics here to be able to run AfterPackage
```
Comment (ajanata) [inline run.go:26]: this whole thing is kinda ugly and I can't think of a way to test it since it is supposed to run inside a test anyway.

File: run.go [hunk 6]
```diff
@@ -13,48 +15,223 @@ import (
...
+		pkgHelperT := &t{}
 
-			wg := sync.WaitGroup{}
-			wg.Add(1)
-			go func() {
-				defer wg.Done()
-				for r := range res {
-					results[pkg][r.Name] = r
-				}
+		// we have to hold onto any panics here to be able to run AfterPackage
```
Comment (ajanata) [inline run.go:100]: and this is unfortunately even uglier, though it is rather extensively tested

File: run.go [hunk 7]
```diff
@@ -13,48 +15,223 @@ import (
...
+		if beforePkgErr == nil && afterPkgErr != nil {
+			newResults := make(orderedmap.OrderedMap[string, TestResult])
+			results[pkg].Iterate(func(test string, result TestResult) bool {
+				newResults[test] = TestResult{
+					Package: result.Package,
+					Name:    result.Name,
+					Passed:  false,
+					Dur:     result.Dur,
+					Msgs: append(result.Msgs, append(pkgHelperT.msgs, Msg{
+						Msg:   fmt.Sprintf("%v", beforePkgErr),
```
Comment (jeffatgametime) [inline run.go]:
```suggestion
						Msg:   fmt.Sprintf("%v", afterPkgErr),
```

but why are we not reporting `afterPkgErr` if `beforePkgErr` is not nil?

File: register.go (hunk 3 above)
Comment (kendallm) [reply register.go:83]: How does this work out if you are testing across flies in the same package.

eg

file1.go
```
package gametime
```

file2.go
```
package gametime
```

I can see use cases where this is desirable and from reading the implementation it does not seem like it would work.

File: run.go (hunk 7 above)
Comment (ajanata) [reply run.go]: can't panic two things, and if before failed any failures in after could very well be related.

Comment (ajanata) [reply run.go]: wait, this isn't the path that panics, it should have already been reported above... though do the tests reflect that? hm.

Comment (ajanata) [reply run.go]: ah, I don't have any test cases with multiple things failing, let me add one. but I think it should have them both

File: run.go [hunk 8]
```diff
@@ -13,48 +15,223 @@ import (
...
-	instance.tests.Iterate(func(pkg string, tests orderedmap.OrderedMap[string, testCase]) bool {
-		tests.Iterate(func(name string, test testCase) bool {
-			t.Run(test.Name, func(tt *testing.T) {
-				tt.Helper()
-				test.tester(tWrapper{t: tt})
+	instance.tests.Iterate(func(pkg string, pkgTests *testPkg) bool {
+		// we have to hold onto any panics here to be able to run AfterPackage
+		var beforePkgErr any
+		if pkgTests.BeforePackage != nil {
+			func() {
```
Comment (kendallm) [inline run.go:29]: Is it necessary to wrap this in an anonymous function here?

File: run.go (hunk 7 above)
Comment (ajanata) [reply run.go]: ok, yeah, this was easy enough to do in Run mode

**Review by kendallm — COMMENTED** [review]
Still wrapping my head around the concurrency stuff, but LGTM based on the testing you have in here. Very helpful!

Couple questions though.

**Review by jeffatgametime — APPROVED** [review]

File: register.go (hunk 3 above)
Comment (ajanata) [reply register.go:83]: If they absolutely need that behavior, they can be in different packages. There isn't an easy way to get more fine grain control in a generic manner. Alternatively (and this is what we do a lot already), you can have common setup in the outer level test and have subtests that do the actual testing. These are designed more for stuff like spinning up containers (which we will need eventually) or creating entries in a database.

File: run.go (hunk 8 above)
Comment (ajanata) [reply run.go:29]: yes, because the `defer` needs to apply only to `pkgTests.BeforePackage`

**Review by kendallm — APPROVED** [review]
